**This notebook is an exercise in the [Introduction to Machine Learning](https://www.kaggle.com/learn/intro-to-machine-learning) course.  You can reference the tutorial at [this link](https://www.kaggle.com/alexisbcook/machine-learning-competitions).**

---


# Introduction

In this exercise, you will create and submit predictions for a Kaggle competition. You can then improve your model (e.g. by adding features) to apply what you've learned and move up the leaderboard.

Begin by running the code cell below to set up code checking and the filepaths for the dataset.

In [2]:
# Set up code checking
from learntools.core import binder
binder.bind(globals())
from learntools.machine_learning.ex7 import *

# Set up filepaths
import os
if not os.path.exists("../input/train.csv"):
    os.symlink("../input/home-data-for-ml-course/train.csv", "../input/train.csv")  
    os.symlink("../input/home-data-for-ml-course/test.csv", "../input/test.csv") 

Here's some of the code you've written so far. Start by running it again.

In [3]:
# Import helpful libraries
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split

# Load the data, and separate the target
iowa_file_path = '../input/train.csv'
home_data = pd.read_csv(iowa_file_path)
y = home_data.SalePrice

# Create X (After completing the exercise, you can return to modify this line!)
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']

# Select columns corresponding to features, and preview the data
X = home_data[features]
X.head()

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Define a random forest model
rf_model = RandomForestRegressor(random_state=1)
rf_model.fit(train_X, train_y)
rf_val_predictions = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

print("Validation MAE for Random Forest Model: {:,.0f}".format(rf_val_mae))

Validation MAE for Random Forest Model: 21,857


Our MAE is 21,857 now.

In [4]:
# adding only one column to the provided list of columns to define the best list of features
def picking_columns(cols):
    features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
    
    for col in cols:
        features.append(col)
        
        X = home_data[features]
        train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)
        
        rf_model = RandomForestRegressor(random_state=1)
        rf_model.fit(train_X, train_y)
        rf_val_predictions = rf_model.predict(val_X)
        rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

        print("Validation MAE for Random Forest Model {}: {:,.0f}".format(col, rf_val_mae))

In [5]:
valid_columns = ['MSSubClass', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt',
                 'YearRemodAdd', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea',
                 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
                 'Fireplaces', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch',
                 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold']

In [6]:
valid_columns_0 = ['MSSubClass', 'OverallQual', 'OverallCond', 'YearRemodAdd', 'LowQualFinSF',
                   'GrLivArea', 'HalfBath', 'KitchenAbvGr', 'Fireplaces', 'WoodDeckSF',
                   'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
                   'MiscVal', 'MoSold', 'YrSold']

In [7]:
picking_columns(valid_columns_0)

Validation MAE for Random Forest Model MSSubClass: 21,675
Validation MAE for Random Forest Model OverallQual: 19,608
Validation MAE for Random Forest Model OverallCond: 18,701
Validation MAE for Random Forest Model YearRemodAdd: 18,681
Validation MAE for Random Forest Model LowQualFinSF: 18,754
Validation MAE for Random Forest Model GrLivArea: 17,841
Validation MAE for Random Forest Model HalfBath: 17,825
Validation MAE for Random Forest Model KitchenAbvGr: 17,759
Validation MAE for Random Forest Model Fireplaces: 17,510
Validation MAE for Random Forest Model WoodDeckSF: 17,438
Validation MAE for Random Forest Model OpenPorchSF: 17,445
Validation MAE for Random Forest Model EnclosedPorch: 17,634
Validation MAE for Random Forest Model 3SsnPorch: 17,468
Validation MAE for Random Forest Model ScreenPorch: 17,434
Validation MAE for Random Forest Model PoolArea: 17,382
Validation MAE for Random Forest Model MiscVal: 17,596
Validation MAE for Random Forest Model MoSold: 17,856
Validation MAE

Our MAE is 17,382 now. That's an improvement.

In [71]:
# going through all columns and trying to find the best combination of features to decrease MAE we have after impementing previous steps.
import numpy as np

def picking_columns_2(cols):
    dct = {}
    best_score = np.inf
    dct[best_score] = []
    result = []
    
    for i in range(len(cols)):
        for col in cols:
            features = result
            if col in features:
                continue
            features.append(col)

            X = home_data[features]
            train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

            rf_model = RandomForestRegressor(random_state=1)
            rf_model.fit(train_X, train_y)
            rf_val_predictions = rf_model.predict(val_X)
            rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

            print(features, round(rf_val_mae, 2), sep=': ')
            
            features.remove(col)

            if rf_val_mae < best_score:
                best_score = rf_val_mae
                best_feature = col

        if best_score in dct:
            pass
        else:
            result.append(best_feature)
            dct[best_score] = result
        print(f'This is result: {result}: {best_score}')
            

    result = dct[best_score]
    print(f'The very final result: {result}: {best_score}')
    return result

In [72]:
best_features = picking_columns_2(valid_columns)

['MSSubClass']: 48346.44
['LotArea']: 56673.17
['OverallQual']: 28745.2
['OverallCond']: 54983.58
['YearBuilt']: 42018.51
['YearRemodAdd']: 46819.13
['1stFlrSF']: 52518.44
['2ndFlrSF']: 50854.45
['LowQualFinSF']: 59008.55
['GrLivArea']: 44888.72
['FullBath']: 47255.66
['HalfBath']: 55356.06
['BedroomAbvGr']: 57455.19
['KitchenAbvGr']: 58146.92
['TotRmsAbvGrd']: 48831.32
['Fireplaces']: 51079.66
['WoodDeckSF']: 55320.24
['OpenPorchSF']: 52921.78
['EnclosedPorch']: 58322.25
['3SsnPorch']: 58975.64
['ScreenPorch']: 60189.01
['PoolArea']: 59215.69
['MiscVal']: 59201.59
['MoSold']: 59722.86
['YrSold']: 59300.3
This is result: ['OverallQual']: 28745.2020686596
['OverallQual', 'MSSubClass']: 27878.06
['OverallQual', 'LotArea']: 30821.47
['OverallQual', 'OverallCond']: 29139.83
['OverallQual', 'YearBuilt']: 29683.78
['OverallQual', 'YearRemodAdd']: 28626.29
['OverallQual', '1stFlrSF']: 31499.97
['OverallQual', '2ndFlrSF']: 28805.33
['OverallQual', 'LowQualFinSF']: 28817.64
['OverallQual', 'GrL

In [73]:
print(best_features)

['OverallQual', 'TotRmsAbvGrd', 'FullBath', '1stFlrSF', 'YearBuilt', '2ndFlrSF', 'OverallCond', 'LotArea', 'GrLivArea', 'Fireplaces', 'YearRemodAdd', 'MSSubClass', 'ScreenPorch', 'WoodDeckSF']


In [74]:
# Select columns corresponding to features, and preview the data
X = home_data[best_features]
X.head()

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Define a random forest model
rf_model = RandomForestRegressor(random_state=1)
rf_model.fit(train_X, train_y)
rf_val_predictions = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

print("Validation MAE for Random Forest Model: {:,.0f}".format(rf_val_mae))

Validation MAE for Random Forest Model: 17,269


Our MAE is 17,269 now. That's an improvement again.

# Train a model for the competition

The code cell above trains a Random Forest model on **`train_X`** and **`train_y`**.  

Use the code cell below to build a Random Forest model and train it on all of **`X`** and **`y`**.

In [75]:
# To improve accuracy, create a new Random Forest model which you will train on all training data
rf_model_on_full_data = RandomForestRegressor(random_state=1)

# fit rf_model_on_full_data on all data from the training data
rf_model_on_full_data.fit(X, y)

RandomForestRegressor(random_state=1)

Now, read the file of "test" data, and apply your model to make predictions.

In [76]:
# path to file you will use for predictions
test_data_path = '../input/test.csv'

# read test data file using pandas
test_data = pd.read_csv(test_data_path)

# create test_X which comes from test_data but includes only the columns you used for prediction.
# The list of columns is stored in a variable called features
test_X = test_data[best_features]

# make predictions which we will submit. 
test_preds = rf_model_on_full_data.predict(test_X)

Before submitting, run a check to make sure your `test_preds` have the right format.

In [78]:
# Check your answer (To get credit for completing the exercise, you must get a "Correct" result!)
#step_1.check()
# step_1.solution()

# Generate a submission

Run the code cell below to generate a CSV file with your predictions that you can use to submit to the competition.

In [79]:
# Run the code to save predictions in the format used for competition scoring

output = pd.DataFrame({'Id': test_data.Id,
                       'SalePrice': test_preds})
output.to_csv('submission.csv', index=False)

# Submit to the competition

To test your results, you'll need to join the competition (if you haven't already).  So open a new window by clicking on **[this link](https://www.kaggle.com/c/home-data-for-ml-course)**.  Then click on the **Join Competition** button.

![join competition image](https://i.imgur.com/axBzctl.png)

Next, follow the instructions below:
1. Begin by clicking on the **Save Version** button in the top right corner of the window.  This will generate a pop-up window.  
2. Ensure that the **Save and Run All** option is selected, and then click on the **Save** button.
3. This generates a window in the bottom left corner of the notebook.  After it has finished running, click on the number to the right of the **Save Version** button.  This pulls up a list of versions on the right of the screen.  Click on the ellipsis **(...)** to the right of the most recent version, and select **Open in Viewer**.  This brings you into view mode of the same page. You will need to scroll down to get back to these instructions.
4. Click on the **Output** tab on the right of the screen.  Then, click on the file you would like to submit, and click on the blue **Submit** button to submit your results to the leaderboard.

You have now successfully submitted to the competition!

If you want to keep working to improve your performance, select the **Edit** button in the top right of the screen. Then you can change your code and repeat the process. There's a lot of room to improve, and you will climb up the leaderboard as you work.


# Continue Your Progress
There are many ways to improve your model, and **experimenting is a great way to learn at this point.**

The best way to improve your model is to add features.  To add more features to the data, revisit the first code cell, and change this line of code to include more column names:
```python
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
```

Some features will cause errors because of issues like missing values or non-numeric data types.  Here is a complete list of potential columns that you might like to use, and that won't throw errors:
- 'MSSubClass'
- 'LotArea'
- 'OverallQual' 
- 'OverallCond' 
- 'YearBuilt'
- 'YearRemodAdd' 
- '1stFlrSF'
- '2ndFlrSF' 
- 'LowQualFinSF' 
- 'GrLivArea'
- 'FullBath'
- 'HalfBath'
- 'BedroomAbvGr' 
- 'KitchenAbvGr' 
- 'TotRmsAbvGrd' 
- 'Fireplaces' 
- 'WoodDeckSF' 
- 'OpenPorchSF'
- 'EnclosedPorch' 
- '3SsnPorch' 
- 'ScreenPorch' 
- 'PoolArea' 
- 'MiscVal' 
- 'MoSold' 
- 'YrSold'

Look at the list of columns and think about what might affect home prices.  To learn more about each of these features, take a look at the data description on the **[competition page](https://www.kaggle.com/c/home-data-for-ml-course/data)**.

After updating the code cell above that defines the features, re-run all of the code cells to evaluate the model and generate a new submission file.  


# What's next?

As mentioned above, some of the features will throw an error if you try to use them to train your model.  The **[Intermediate Machine Learning](https://www.kaggle.com/learn/intermediate-machine-learning)** course will teach you how to handle these types of features. You will also learn to use **xgboost**, a technique giving even better accuracy than Random Forest.

The **[Pandas](https://kaggle.com/Learn/Pandas)** course will give you the data manipulation skills to quickly go from conceptual idea to implementation in your data science projects. 

You are also ready for the **[Deep Learning](https://kaggle.com/Learn/intro-to-Deep-Learning)** course, where you will build models with better-than-human level performance at computer vision tasks.

---




*Have questions or comments? Visit the [Learn Discussion forum](https://www.kaggle.com/learn-forum/161285) to chat with other Learners.*